Add gotcha to RDoc of collection_check_boxes

maurogeorge opened the pull request:
The `collection_check_boxes` generates a [default hidden input](https://github.com/rails/rails/blob/master/actionview/test/template/form_collections_helper_test.rb#L210-L236) to prevent a gotcha, like the [file_field](https://github.com/rails/rails/blob/master/actionview/lib/action_view/helpers/form_helper.rb#L864-L880) does.

This PR add a RDoc to this.

maurogeorge committed:
Add gotcha to RDoc of collection_check_boxes

[ci skip]

zzak commented on actionview/lib/action_view/helpers/form_options_helper.rb:
```diff
@@ -707,6 +707,24 @@ def collection_radio_buttons(object, method, collection, value_method, text_meth
+      # ==== Gotcha
+      #
+      # The HTML specification says when nothing is select on a collection of radio buttons
+      # web browsers do not send any value to server.
```
Which spec?

Also, probably:
> The HTML specification says that web browsers should not send any value when no selection is made for a collection of radio buttons.

or

> When no selection is made for a collection of radio buttons most web browsers will not send any value, based on the HTML specification.

zzak commented on actionview/lib/action_view/helpers/form_options_helper.rb:
```diff
@@ -707,6 +707,24 @@ def collection_radio_buttons(object, method, collection, value_method, text_meth
+      #
+      # The HTML specification says when nothing is select on a collection of radio buttons
+      # web browsers do not send any value to server.
+      # Unfortunately this introduces a gotcha:
```
I think this is implicit

zzak commented on actionview/lib/action_view/helpers/form_options_helper.rb:
```diff
@@ -707,6 +707,24 @@ def collection_radio_buttons(object, method, collection, value_method, text_meth
+      # web browsers do not send any value to server.
+      # Unfortunately this introduces a gotcha:
+      # if a +User+ model has a +category_ids+ field, and in the form none category is selected no +category_ids+ parameter is sent. So,
+      # any mass-assignment idiom like
```
For example, if we have a +User+ model with +category_ids+ field and we have the following code in our update action:

  @user.update(params[:user])

If no +category_ids+ are selected then we can safely assume this field will not be updated.

This is possible thanks to a hidden field generated by the helper method for every collection of radio buttons. This hidden field is given the same field name as the radio buttons with a blank value.

In the rare case you don't want this hidden field, you can pass the <tt>include_hidden: false</tt> option to the helper method.

maurogeorge commented:
@zzak thanks for the review.
I updated the doc following your observations.

zzak commented:
Please wrap it at 80 chars, we may possibly want to consider the following:

From 8.1.2.4 from [RFC1866](https://tools.ietf.org/html/rfc1866#section-8.1.2.4)
> If none of the <INPUT> elements of a set of radio buttons specifies
> `CHECKED', then the user agent must check the first radio button of
>  the set initially.

maurogeorge commented:
@zzak I updated to wrap it on 80 chars.

Actually we are adding docs about collection of checkboxes and not radio buttons, yeah on my first commit this was sent as radio buttons, my mistake sorry :disappointed: .

As you can see on on [8.1.2.3 from RFC1866](https://tools.ietf.org/html/rfc1866#section-8.1.2.3) we do not have the same doc of the radio buttons. So I rewrote the first paragraph.

This makes senses to you? Or you think we have some RFC that describes this behavior?

zzak commented:
This looks good to me, thanks for your contribution @maurogeorge!

maurogeorge commented:
Thanks @zzak for the review and merge
